Solutions/Lookout Cloud Security/Data Connectors/LookoutCSConnector/LookoutCloudSecurityEvents/__init__.py

```diff
@@ -0,0 +1,248 @@
+        bytes_to_hash = bytes(string_to_hash, encoding="utf-8")
+        decoded_key = base64.b64decode(self.sharedkey)
+        encoded_hash = base64.b64encode(hmac.new(decoded_key, bytes_to_hash, digestmod=hashlib.sha256).digest()).decode()
+        authorization = "SharedKey {}:{}".format(customer_id, encoded_hash)
+        return authorization
+
+    def post_data(self, body, chunk_count):
+        method = 'POST'
+        content_type = 'application/json'
+        resource = '/api/logs'
+        rfc1123date = datetime.datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT')
+        content_length = len(body)
+        signature = self.build_signature(rfc1123date, content_length, method, content_type,
+                                         resource)
+        uri = self.logAnalyticsUri + resource + '?api-version=2016-04-01'
+        headers = {
+            'content-type': content_type,
+            'Authorization': signature,
+            'Log-Type': self.table_name,
+            'x-ms-date': rfc1123date
+        }
+        response = requests.post(uri, data=body, headers=headers)
+        if (response.status_code >= 200 and response.status_code <= 299):
+            logging.info("Chunk was processed{} events".format(chunk_count))
+            self.success_processed = self.success_processed + chunk_count
+        else:
+            logging.error("Error during sending events to Microsoft Sentinel. Response code:{}".format(response.status_code))
+            self.fail_processed = self.fail_processed + chunk_count  
+
+# this function app is fired based on the Timer trigger
+# it is used to capture all the events from LookOut cloud security API   
+def main(mytimer: func.TimerRequest) -> None:
+    utc_timestamp = datetime.datetime.utcnow().replace(
+        tzinfo=datetime.timezone.utc).isoformat()
+    if mytimer.past_due:
+     logging.info('The timer is past due!')
+    logging.info('Python timer trigger function ran at %s', utc_timestamp)
+    logging.info('Starting program')
+    results_events = []
+    try:
+        Lookout = LookOut()
+        sentinel = Sentinel()
+        sentinel.sharedkey = shared_key
+        sentinel.table_name= table_name    
+        startTime,endTime = Lookout.generate_date()  
+        logging.info("The current run Start time {}".format(startTime))
+        logging.info("The current run End time {}".format(endTime))        
+        logging.info('Start: to get Activity')
+        results_events = Lookout.get_Data("/apigw/v1/events?eventType=Activity",startTime,endTime)
+        logging.info("The number of Activities processed {} ".format(len(results_events)))
+        logging.info('End: to get Activity')         
+        
+        if(len(results_events)) > 0:
+         # Sort the json based on the "timestamp" key
+         sorted_data = sorted(results_events, key=lambda x: x["timeStamp"],reverse=False) 
+         # Fetch the latest timestamp
+         latest_timestamp = sorted_data[-1]["timeStamp"]       
```

Comment:
Not considering the scenario, what happens when we are able to send only 1-2 batches gen_chunks()  and remaining got failures. We should increment, only when we are able to send data successfully to sentinel.

Reply:
Rewrote the entire chunk process in order to capture the latest timestamp and tested. Please review.